### PR TITLE
BUG: inconsistency between PeriodIndex.get_value vs get_loc

### PR DESCRIPTION
**Pull request by jbrockmendel**
ATM we have inconsistent treatment for indexing with strings where the string has a higher resolution than the PeriodIndex.  Luckily only one test checks for the inconsistent `PeriodIndex.__contains__` behavior, so fixing it isn't _too_ invasive.

```
pi = pd.period_range("2017-09-01", freq="D", periods=3)
ser = pd.Series(range(5, 8), index=pi)
key = "2017-09-01 00:00:01"

# master
>>> key in pi
True
>>> pi.get_loc(key)
0
>>> pi.get_value(ser, key)
KeyError: '2017-09-01 00:00:01'

# PR
>>> key in pi
False
>>> pi.get_loc(key)
KeyError: '2017-09-01 00:00:01'
>>> pi.get_value(ser, key)
KeyError: '2017-09-01 00:00:01'
```

Note: the diff is fairly big because we're mirroring the code in `get_value`.  Once this is fixed, we'll have get_value dispatch to get_loc and remove basically the same amount of code as we're adding here.

### CODE REVIEW

**Review by gfyoung — APPROVED** [review]
> Once this is fixed, we'll have get_value dispatch to get_loc and remove basically the same amount of code as we're adding here.

Sounds like a good follow-up once this is merged.

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -567,17 +567,32 @@ def get_loc(self, key, method=None, tolerance=None):
...
 
             try:
-                asdt, reso = parse_time_string(key, self.freq)
-                key = asdt
-            except DateParseError:
-                # A string with invalid format
-                raise KeyError(f"Cannot interpret '{key}' as period")
+                loc = self._get_string_slice(key)
+                return loc
+            except (TypeError, ValueError):
```
Comment (jreback) [inline pandas/core/indexes/period.py]: OverflowError can no longer occur?

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -567,17 +567,32 @@ def get_loc(self, key, method=None, tolerance=None):
...
+
```
Comment (jreback) [inline pandas/core/indexes/period.py]: can you just pass in the except and out-dent this while block (the _get_string_slice try/except)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/period.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: sure, either way works

File: pandas/core/indexes/period.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: ill see if i can cook up a test case that overflows

**Comment by jbrockmendel** [comment]
updated+green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**